HW2/P4/driver.py

```diff
@@ -19,14 +19,62 @@ def py_median_3x3(image, iterations=10, num_threads=1):
     ''' repeatedly filter with a 3x3 median '''
     tmpA = image.copy()
     tmpB = np.empty_like(tmpA)
-
-    for i in range(iterations):
-        filtering.median_3x3(tmpA, tmpB, 0, 1)
-        # swap direction of filtering
-        tmpA, tmpB = tmpB, tmpA
-
+    locks = [None]*num_threads
+    calc_count = [0]*num_threads
+    swap_count = [0]*num_threads
+    threads = [None]*num_threads
+    for i in range(num_threads):
+        locks[i] = threading.Condition()
+    
+    for i in range(num_threads):
+        threads[i] = threading.Thread(target=median_worker, args=(i, iterations, tmpA, tmpB, locks, calc_count, swap_count))
+        threads[i].start()
+        
+    for t in threads:
+        t.join()
+        
     return tmpA
 
+def median_worker(thread_num, iterations, tmpA, tmpB, locks, calc_count, swap_count):
+    neighbor_threads = set([thread_num-1 if thread_num > 0 else len(locks)-1, thread_num + 1 if thread_num < len(locks) - 1 else 0])
+    for i in range(iterations):  
+        #can't calculate until neighbor threads have been copied to the right iteration
+        #if they haven't been copied yet, wait on corresponding locks until notified
```

Comment:
Nice job with your style and commenting. It really makes it easy to follow along.